### PR TITLE
Rewrote FormattingHelper from scratch. Renamed it. Added tests.

### PR DESCRIPTION
**Pull request by lewisbar**
I rewrote FormattingHelper from scratch using TDD, which resulted in what I think is much better code, and a lot of bugs eliminated on the way.

### CODE REVIEW

**Comment by lewisbar** [comment]
I'm not sure about the name "Completer", but it seems more accurate than "FormattingHelper".

**Comment by lewisbar** [comment]
I just thought I should add tests for the case when the TextView is empty. I'll do it when I find the time.

File: GetSchwifty/Code Helper/Completer.swift [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+//  Completer.swift
+//  GetSchwifty
+//
+//  Created by Lennart Wisbar on 03.05.17.
+//  Copyright © 2017 Benjamin Herzog. All rights reserved.
+//
+
+import Foundation
+
+struct Completer {
```
Comment (BenchR267) [inline GetSchwifty/Code Helper/Completer.swift:11]: This sounds a bit like auto completion 😆 Next feature maybe? 🙈

File: GetSchwifty/Code Helper/Completer.swift [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+    }
+    
+    static func completedInput(for input: String, scenario: Scenario, indentation: Int) -> (String, cursorOffset: Int) {
+        var insertion = input
+        var cursorOffset = input.utf16.count
+        
+        switch scenario {
+        case .normal:
+            insertion = input
+            cursorOffset = insertion.utf16.count
```
Comment (BenchR267) [inline GetSchwifty/Code Helper/Completer.swift:70]: Is utf16 the correct one to use? Why not characters (or in Swift 4 the String directly)?

File: Get SchwiftyTests/CompleterTests.swift [hunk 3]
```diff
@@ -0,0 +1,2477 @@
...
+        let expectedText = "tes a t"
+        let expectedRange = NSMakeRange(5, 0) // After the "a"
+        
+        XCTAssertEqual(newText, expectedText)
+        XCTAssertEqual(newRange.location, expectedRange.location)
+        XCTAssertEqual(newRange.length, expectedRange.length)
+    }
+    
+    func test_NormalCharacter_WithEmojis_InsertedNormally() {
+        let text = "👱🏾🚀test"
```
Comment (BenchR267) [inline Get SchwiftyTests/CompleterTests.swift:106]: 🤣

**Review by BenchR267 — APPROVED** [review]
Looks really awesome! Great contribution! 💪

File: GetSchwifty/Code Helper/Completer.swift (hunk 1 above)
Comment (lewisbar) [reply GetSchwifty/Code Helper/Completer.swift:11]: Yes, that's exactly the problem with this name. Still I thought it fits better than FormattingHelper. I don't know.

File: GetSchwifty/Code Helper/Completer.swift (hunk 2 above)
Comment (lewisbar) [reply GetSchwifty/Code Helper/Completer.swift:70]: Yes. I used characters first, but that doesn't work correctly with emojis. This method returns a cursorOffset which is used to determine the new cursor position in the UITextView. The text view (i.e. the method shouldChangeText) works with NSRange, not with String.Index. I'm sure utf16 is what we need here.

File: Get SchwiftyTests/CompleterTests.swift (hunk 3 above)
Comment (lewisbar) [reply Get SchwiftyTests/CompleterTests.swift:106]: 🤡 This is a good example why we need utf16. With characters, the cursor position would be off after entering this text.

File: GetSchwifty/Code Helper/Completer.swift (hunk 1 above)
Comment (BenchR267) [reply GetSchwifty/Code Helper/Completer.swift:11]: Sure, it's not a problem until we introduce autocompletion :)

File: GetSchwifty/Code Helper/Completer.swift (hunk 2 above)
Comment (BenchR267) [reply GetSchwifty/Code Helper/Completer.swift:70]: Mh crazy, I thought a character is exactly one unicode symbol, so also merged emojis like 🙎🏾. Maybe this is one of the changes in Swift 4

**Review by BenchR267 — DISMISSED** [review]

File: GetSchwifty/Code Helper/Completer.swift (hunk 2 above)
Comment (lewisbar) [reply GetSchwifty/Code Helper/Completer.swift:70]: Yes, `characters` should generally count an emoji as one character*. But to communicate with the UITextView, we need to use NSRange. In the UITextView, emojis can have a longer range (NSRange) than 1. So the code needs to translate between `NSRange` and `Range<String.Index>`. Otherwise, the cursor position will be off and you could theoretically end up in the middle of an emoji.
At first, I had used `characters` instead of `utf16`, and this led to chaos as soon as emojis were involved. With `utf16`, those problems are fixed, as expected.
I tried to keep the String extension free of NSRange and do the translation between the two systems only in the Completer struct, in an effort for clean code.

*Although that's not even always true: https://oleb.net/blog/2016/08/swift-3-strings/

**Comment by lewisbar** [comment]
Like I mentioned earlier, I thought tests for empty text were missing. So I added them.

File: GetSchwifty/Code Helper/Completer.swift (hunk 2 above)
Comment (lewisbar) [reply GetSchwifty/Code Helper/Completer.swift:70]: I need to correct myself: "I tried to keep the String extension free of NSRange and do the translation between the two systems only in the Completer struct" - That's not entirely true. Of course, the `func stringRange(from range: NSRange) -> Range<String.Index>?` belongs in the String extension, as it needs to work with a certain String to be able to return a `Range<String.Index>`. At least in my opinion ...

Comment (BenchR267) [reply GetSchwifty/Code Helper/Completer.swift:70]: If you would have said to me that we are discussing about emojis in 2017 5 years ago I would had laughed about you :D

**Review by BenchR267 — APPROVED** [review]
Looks still awesome, you are free to merge it! 💪

**Comment by lewisbar** [comment]
👍

File: Get SchwiftyTests/CompleterTests.swift (hunk 3 above)
Comment (uliwitness) [reply Get SchwiftyTests/CompleterTests.swift:106]: Not fully following this convo, but to test whether you're doing Unicode correctly, be sure to test with some of the following:

1. A country flag. Those are actually two distinct characters ("C" and "A" for Canada, but using a special country flag emoji alphabet)
2. An Emoji with a skin tone modifier (I think that one was a surrogate pair)
3. Any other "regular Emoji" (those are single code points IIRC).

It's been a while since I had to muck with Unicode at the byte level, but definitely be aware of multiple "characters" for one character (even in UTF-32), and the case of both ones that can be detected as being one character (like combining accents on un-accented letters) and ones that are two distinct characters but produce a single glyph (like country flags).

Comment (lewisbar) [reply Get SchwiftyTests/CompleterTests.swift:106]: @uliwitness Thank you for your thoughts on this.
So you think country flags are significantly different from emojis with a skin tone modifier? I could, of course, add some tests for country flags. The flags I tried out in the playground all had a utf16 count of 4.
As to your 3rd point: I tried to find an emoji that only has a utf16 count of 1 in the playground, but couldn't find one. Even the simplest emojis like 🙂 and 👍 seem to count at least 2.
